src/parser/classish.rs

```diff
@@ -0,0 +1,254 @@
+            extends = Some(self.full_name()?.into());
+        }
+
+        let mut implements = Vec::new();
+        if self.current.kind == TokenKind::Implements {
+            self.next();
+
+            while self.current.kind != TokenKind::LeftBrace {
+                self.optional_comma()?;
+
+                implements.push(self.full_name()?.into());
+            }
+        }
+
+        self.lbrace()?;
+
+        let mut body = Vec::new();
+        while self.current.kind != TokenKind::RightBrace && !self.is_eof() {
+            body.push(self.anonymous_class_statement()?);
+        }
+
+        self.rbrace()?;
+
+        Ok(Expression::New {
+            target: Box::new(Expression::AnonymousClass {
+                extends,
+                implements,
+                body,
+            }),
+            args,
+        })
+    }
+
+    pub(crate) fn enum_definition(&mut self) -> ParseResult<Statement> {
+        self.next();
+
+        let name = self.ident()?;
+
+        let backed_type: Option<BackedEnumType> = if self.current.kind == TokenKind::Colon {
+            self.colon()?;
+
+            match self.current.kind.clone() {
+                TokenKind::Identifier(s) if s == b"string" || s == b"int" => {
+                    self.next();
+
+                    Some(match &s[..] {
+                        b"string" => BackedEnumType::String,
+                        b"int" => BackedEnumType::Int,
+                        _ => unreachable!(),
+                    })
+                }
+                _ => {
+                    return expected_token_err!(["`string`", "`int`"], self);
+                }
+            }
+        } else {
+            None
+        };
+
+        let mut implements = Vec::new();
+        if self.current.kind == TokenKind::Implements {
+            self.next();
+
+            while self.current.kind != TokenKind::LeftBrace {
+                implements.push(self.full_name()?.into());
+
+                self.optional_comma()?;
+            }
+        }
+
+        self.lbrace()?;
+
+        let mut body = Block::new();
+        while self.current.kind != TokenKind::RightBrace {
+            self.skip_comments();
+            body.push(self.enum_statement(backed_type.is_some())?);
+        }
+
+        self.rbrace()?;
+
+        match backed_type {
+            Some(backed_type) => Ok(Statement::BackedEnum {
+                name: name.into(),
+                backed_type,
+                implements,
+                body,
+            }),
+            None => Ok(Statement::UnitEnum {
+                name: name.into(),
+                implements,
+                body,
+            }),
+        }
+    }
+
+    fn at_least_one_comma_seperated<T>(
+        &mut self,
+        func: &(dyn Fn(&mut Parser) -> ParseResult<T>),
+    ) -> ParseResult<Vec<T>> {
+        let mut result: Vec<T> = vec![];
+        loop {
+            result.push(func(self)?);
+            if self.current.kind == TokenKind::Comma {
```

Comment:
Flip these round so we test unhappy first, i.e. `self.current.kind != TokenKind::Comma`